### PR TITLE
adding retry functionality and batch processing

### PR DESCRIPTION
**Pull request by ssaini-phyllo**

### CODE REVIEW

File: rabbitmq_client/publisher.py [hunk 1]
```diff
@@ -1,12 +1,14 @@
 import json
+import logging
 from datetime import datetime, date
 from uuid import UUID
 
-import pika
+import pika.exceptions
```
Comment (anujith-singh) [inline rabbitmq_client/publisher.py]: is this change required?

File: rabbitmq_client/publisher.py [hunk 2]
```diff
@@ -34,27 +36,59 @@ def register_serializer(self, func: Callable):
             raise DuplicateSerializerError(f"Serializer already assigned with name '{self.serializer.__name__}'")
         self.serializer = func
 
-    def publish(self, queue_config: PublishQueueConfig, payload: dict, headers: Optional[dict] = None,
-                priority: Optional[int] = 0):
+    def publish(self, queue_config: PublishQueueConfig,
+                payload: Optional[dict] = None, payloadList: List[dict] = None,
```
Comment (anujith-singh) [inline rabbitmq_client/publisher.py]:
```suggestion
                payload: Optional[dict] = None, payload_list: List[dict] = None,
```

File: rabbitmq_client/publisher.py [hunk 3]
```diff
@@ -34,27 +36,59 @@ def register_serializer(self, func: Callable):
             raise DuplicateSerializerError(f"Serializer already assigned with name '{self.serializer.__name__}'")
         self.serializer = func
 
-    def publish(self, queue_config: PublishQueueConfig, payload: dict, headers: Optional[dict] = None,
-                priority: Optional[int] = 0):
```
Comment (anujith-singh) [inline rabbitmq_client/publisher.py:38]: are we deprecating `priority`?

File: rabbitmq_client/publisher.py [hunk 4]
```diff
@@ -34,27 +36,59 @@ def register_serializer(self, func: Callable):
...
         if headers is None:
             headers = {}
         assert isinstance(queue_config, PublishQueueConfig), \
             f"Expected instance of PublishQueueConfig, passed {type(queue_config)}"
-        assert isinstance(payload, dict), \
-            f"Expected instance of dict, passed {type(payload)}"
-        assert isinstance(headers, dict), \
-            f"Expected instance of dict, passed {type(headers)}"
-        assert isinstance(priority, int), \
-            f"Expected instance of int, passed {type(priority)}"
```
Comment (anujith-singh) [inline rabbitmq_client/publisher.py:48]: intended change?

Comment (ssaini-phyllo) [reply rabbitmq_client/publisher.py:48]: reverted

File: rabbitmq_client/publisher.py (hunk 3 above)
Comment (ssaini-phyllo) [reply rabbitmq_client/publisher.py:38]: reverted

File: rabbitmq_client/publisher.py [hunk 5]
```diff
@@ -34,27 +36,51 @@ def register_serializer(self, func: Callable):
...
         if headers is None:
             headers = {}
         assert isinstance(queue_config, PublishQueueConfig), \
             f"Expected instance of PublishQueueConfig, passed {type(queue_config)}"
-        assert isinstance(payload, dict), \
-            f"Expected instance of dict, passed {type(payload)}"
-        assert isinstance(headers, dict), \
-            f"Expected instance of dict, passed {type(headers)}"
+        assert self.serializer is not None, \
+            "Define custom serializer with @publish.register_serializer decorator"
```
Comment (anujith-singh) [inline rabbitmq_client/publisher.py]: 🤔

File: rabbitmq_client/publisher.py [hunk 6]
```diff
@@ -34,27 +36,51 @@ def register_serializer(self, func: Callable):
...
+        assert isinstance(headers, dict), f"Expected instance of dict, passed {type(headers)}"
         assert isinstance(priority, int), \
             f"Expected instance of int, passed {type(priority)}"
 
-        serializer: Callable = self.serializer or default_serializer
```
Comment (anujith-singh) [inline rabbitmq_client/publisher.py:50]: deprecating `default_serializer`?

File: rabbitmq_client/publisher.py [hunk 7]
```diff
@@ -34,27 +36,51 @@ def register_serializer(self, func: Callable):
...
+        # Check that either payload or payload_list is provided, but not both None
+        if payload is None and (payload_list is None or len(payload_list) == 0):
+            raise ValueError("Either 'payload' or 'payload_list' must be provided.")
+
+        if payload_list is None or len(payload_list) == 0:
+            payload_list = [payload]
+
+        for payload_item in payload_list:
+            assert isinstance(payload_item, dict), \
+                f"Expected instance of dict, passed {type(payload_item)}"
```
Comment (anujith-singh) [inline rabbitmq_client/publisher.py]: can this be done at the time of publish?
so that we dont' have to run the loop twice

File: rabbitmq_client/publisher.py [hunk 8]
```diff
@@ -34,27 +36,51 @@ def register_serializer(self, func: Callable):
...
+                # channel.queue_declare(queue=queue_config.name)
+                try:
+                    channel.basic_publish(exchange=queue_config.exchange,
+                                          routing_key=queue_config.routing_key,
+                                          body=stringified_payload,
+                                          properties=pika.BasicProperties(headers=headers, priority=priority))
+                except Exception as e:
+                    logging.error(f"Got exception {e} while publishing message to queue {queue_config.routing_key}")
+                    raise PublisherException(f"Got exception {e} while publishing message to queue "
+                                             f"{queue_config.routing_key} for payload {payload}")
```
Comment (anujith-singh) [inline rabbitmq_client/publisher.py]: if the idea is to log and reraise, then can do like this?
```suggestion
                    raise
```

File: rabbitmq_client/publisher.py [hunk 9]
```diff
@@ -34,27 +36,51 @@ def register_serializer(self, func: Callable):
...
+        if payload_list is None or len(payload_list) == 0:
+            payload_list = [payload]
+
+        for payload_item in payload_list:
+            assert isinstance(payload_item, dict), \
+                f"Expected instance of dict, passed {type(payload_item)}"
+
+        self._publish_to_channel(queue_config, payload_list, headers, priority)
+
+    def _publish_to_channel(self, queue_config: PublishQueueConfig, payloads: List[dict],
```
Comment (anujith-singh) [inline rabbitmq_client/publisher.py]:
```suggestion
    def _atomic_publish_to_channel(self, queue_config: PublishQueueConfig, payloads: List[dict],
```

File: rabbitmq_client/publisher.py (hunk 6 above)
Comment (ssaini-phyllo) [reply rabbitmq_client/publisher.py:50]: reverted

File: rabbitmq_client/publisher.py (hunk 7 above)
Comment (ssaini-phyllo) [reply rabbitmq_client/publisher.py]: Ideally we should do validations separately from business logic.

File: rabbitmq_client/publisher.py (hunk 9 above)
Comment (ssaini-phyllo) [reply rabbitmq_client/publisher.py]: renamed

File: rabbitmq_client/publisher.py (hunk 8 above)
Comment (ssaini-phyllo) [reply rabbitmq_client/publisher.py]: changed

File: rabbitmq_client/publisher.py (hunk 5 above)
Comment (ssaini-phyllo) [reply rabbitmq_client/publisher.py]: removed validations on serializer

**Review by anujith-singh — APPROVED** [review]

**Review by pranav-phyllo — APPROVED** [review]
